### ISSUE TITLE
3.3.0 release planning

### ISSUE THREAD

**Issue by mbest** [knockout/knockout#1673]
I'd like to get 3.3.0 out the door this month if possible. That probably means that most of the currently 22 open issues in the 3.3.0 milestone will have to get pushed back.

Here are the ones that I want to get in:

* [x] #1543 Add caching to sleeping pure computed observables (I've been working on this since September.)
* [x] #1427 Make ko.dependencyDetection.ignore exported outside (We just need to figure out the name.)
* [x] #1526 Don't use AMD-style `require` in CommonJS environment

The rest I think we can safely delay until 3.4.0 or later.

@rniemeyer @SteveSanderson @brianmhunt - Thoughts?

**Comment by VoiceOfWisdom** [comment]
Any chance of getting #1653 in as well? Its a regression in 3.2 that is affecting some of my projects. I have a custom build of knockout in production including that PR, and everything is working.

**Comment by brianmhunt** [comment]
Thanks @mbest - I think this is a good short-list, and it would be good to not delay release.

If there were a chance for #1628, I would be mighty happy, but would prefer it not delay the release.

On #1427, what about `ko.peek(callback, callbackTarget, callbackArgs)`? Just a thought.

On an aside, and not to get ahead of ourselves, are there any objections to making the build system in #1360 / #1039 the main (only) issue for 3.4, and push the balance of the issues off to 3.5? It would be great to get that off the table - once in there, it should speed up developing and testing substantially.

**Comment by rniemeyer** [comment]
@brianmhunt - For 3.4, I don't think we would want a release that only updates our build system. Fine to look at getting it in there soon.

**Comment by brianmhunt** [comment]
@rniemeyer No problem!

It feels important to momentarily pause other development until the new build system is hammered out, so we do not miss any changes that need to be duplicated manually in the gulp build (which are generally in the form of new files, and I have been doing separately until now). A milestone was just the first way that came to mind to do that.

**Comment by mbest** [comment]
@VoiceOfWisdom, as I mentioned in that issue, I think using `ko.pureComputed` along with the technique I showed in #1576 will give you what you need.

**Comment by mbest** [comment]
Just looking briefly at #1628, I think it will take some time to get it ready to merge and that we should wait until after 3.3.0.

**Comment by brianmhunt** [comment]
Thanks @mbest; much appreciated. After 3.3 it is. :smile:

**Comment by SteveSanderson** [comment]
Excellent - thanks for pushing this forwards! Totally agreed that we can get this release out very soon. I think all three of the items on your list can be merged/implemented very soon - just need the ongoing discussions to come to a close.

After that there's the docs to update, of course :)

**Comment by brianmhunt** [comment]
I trust nobody will mind, but I will go ahead and mass move the 3.3 milestone items not listed above to a new 3.4.0 milestone. I'm hoping it'll make 3.3 feel closer. :)

**Comment by SteveSanderson** [comment]
3.3 does indeed feel very close, since the only remaining item in the milestone is this release planning issue :)

But of course there's docs. Here's what I think we need to update:

- [ ] **Pure computeds:** New `awake`/`asleep` notifications, clarification on which kinds of subscriptions cause wake/sleep, update any descriptions of pureComputed behaviors that are affected by caching (I think there were some warnings about possible perf drawbacks of pureComputed that no longer apply)
- [ ] **Computeds:** `getSubscriptionsCount` now takes an option `event` param. Newly-public `ko.ignoreDependencies` API
- [ ] **Components:** synchronous DOM injection, receiving `templateNodes` as a property on `componentInfo`, passing bindable markup (`$componentTemplateNodes`), `$component` binding context variable
- [ ] **Template binding:** New `nodes` option
- [ ] **Improved cross-window support** (not sure if this should be in docs really, or where we'd write about it - might need clarification on exactly what scenarios we want to consider officially supported)
- [ ] Consider adding a **new doc page showing SVG examples**, since such a lot of work has gone into making SVG work well with KO bindings
- [ ] Also write **release notes**. This will be a list of the features mentioned here, plus a list of all the bugfixes merged since 3.2.0.

I'd be keen to write all the updated "components" docs. Any takers for other items?

**Comment by brianmhunt** [comment]
Great list. It would be nice to have `trackArrayChanges` in the docs, too (seems conspicuously absent).

**Comment by rniemeyer** [comment]
I pushed a `gh-pages-3.3.0` branch and did a couple of the doc items. I'll try to scrape together a preliminary set of release notes too.

**Comment by mbest** [comment]
I'll work on the pure computed changes.

**Comment by rniemeyer** [comment]
Here are some preliminary release notes:

Enhancements
------------
#1504 - Component configuration can specify a `synchronous` flag to render already loaded components synchronously.
#1463 - Component child elements are provided to `createViewModel` method, component child elements are exposed on binding context as `$componentTemplateNodes`, and `template` binding can accept an array of DOM nodes directly through the `nodes` option.
#1449 - Binding context now contains a `$component` property that references nearest component view model (allows easily binding to the root of the current component).
#1596 - Create a writable computed within a custom element component when an expression evaluates to an observable (less need to use `.$raw` property).
#1576 - computed observables (pure or deferred) notify an `awake` event with the current value when awakening, pure computeds notify an `asleep` event with an `undefined` value when going to sleep, and the `getSubscriptionsCount` method on a computed can accept an event name argument to return only the count of subscriptions for that event name.
#1543 - Sleeping pure computed observables now cache their value
#1635 - Export `ko.utils.setTextContent`
#1427 - Export `ko.dependencyDetection.ignore` as `ko.ignoreDependencies`

Fixes
-----
#1321 - Ensure that template nodes are created and moved within the correct document to fix cross-window issues in IE.
#1434 - Binding parser skips all colons between the name and value. Support returning malformed bindings (non key-value pairs).
#1468 - CSS binding was not clearing a previously added dynamic class when the value bound against changes to null.
#1507 - Memory Leak fix for foreach/template binding
#1513 - Fix error in `fixUpContinuousNodeArray` when using `jquery.tmpl`
#1515 - Update `dependentObservable._latestValue` when a pureComputeds wakes up (affects debug build only)
#1526 - Component's `require` option should not try to do an AMD-style `require` in a CommonJS scenario.
#1556 - Check that the value passed to `optionsAfterRender` is a function before calling
#1597 - CSS binding now works properly with SVG elements
#1606 - Don't bind contents of `textarea` elements
#1623 - `valueAllowUnset` - `null` selection should be maintained when `option` items are updated.
#1634 - Better clean-up of `arrayChange` subscriptions
 
I am leaving on a vacation tomorrow without a laptop until 1/31. I would love to write a blog post about 3.3 though when I get back and will help with any remaining tasks.

**Comment by brianmhunt** [comment]
What a great list - thank you @rniemeyer

**Comment by SteveSanderson** [comment]
New component docs are in.

**Comment by dmr** [comment]
Do I read correctly that 3.3.0 is finished feature-wise but docs are missing for a release?

**Comment by nmehlei** [comment]
I'd like to know that myself, @SteveSanderson @rniemeyer could you clarify?

**Comment by mbest** [comment]
@SteveSanderson I know you usually run the test suite on the set of supported browsers and maybe run some other tests. Have you done that already?

**Comment by SteveSanderson** [comment]
I've now run through all the tests on all the browsers. The issues were:

 * Various problems on IE6/7/8 which I've fixed (minor spec issues only, so I pushed the fixes to `master` - no changes to product code needed).
 * Our `Cross window support` tests don't work on old Firefox (I tried v3.5) or on iOS Safari. For mobiles I'm not really surprised, as popups aren't quite the same thing as they are on desktop browsers. For old Firefox I don't know why this is and haven't investigated - anyone want to take a look? It does work on current Firefox. I'm OK with just saying the new cross-window support is not compatible with ancient builds of Firefox.
  * We should amend the test runner so that these particular tests are skipped on inapplicable browsers, so we don't have to remember this caveat during every release
 * On old Firefox (3.5), the test `Should be able to apply the numeric value zero to a style` fails saying `Expected '0pt' to be '0px'`. Clearly this is just a test issue, so need not block the release. I haven't fixed this test as I was focusing on IE things - anyone want to make this pass?

So I think we are good to go with the release! There are a couple of doc items on the checklist left, but we will only wait for those if someone says they are doing them, as they aren't strictly necessary.

**Comment by rniemeyer** [comment]
Great @SteveSanderson . I started a small blog post highlighting some of the features that I think are interesting. I'll try to finish that tonight or tomorrow.

**Comment by dmr** [comment]
Thank you for the clarification :+1:

**Comment by claudeduguay** [comment]
A week ago it sounded like this release was imminent and depended only on a blog entry that could have been out 5 days ago. Any word on the status of this 3.3.0 release? What's still missing?

**Comment by rniemeyer** [comment]
As far as I am concerned, it is ready to go. We just need to tag with the dist files built, add the final files/links into the docs, merge the 3.3 docs into `gh-pages`, and add the release/release notes on GitHub. Anyone else have anything they are waiting on @SteveSanderson, @mbest , @brianmhunt?

I can take care of all of those steps, if we are good to go. The only thing that I would need an assist on is that I don't have rights to publish to NPM.

**Comment by rniemeyer** [comment]
The steps above would include incrementing the version in bower.json and package.json, of course.

**Comment by mbest** [comment]
@rniemeyer I have the NPM rights, so I went ahead and published there. I also updated the the repository with the 3.3.0 build. I'll add the release notes as well.

**Comment by rniemeyer** [comment]
@mbest - I'll update/merge the docs

**Comment by brianmhunt** [comment]
Nothing to add. Cheers! :beers:

**Comment by rniemeyer** [comment]
Ok- 3.3.0 is done!

**Comment by PaulHuizer** [comment]
Horay!!!

**Comment by SteveSanderson** [comment]
Great job with the release! Thanks very much for updating everything!

**Comment by mbest** [comment]
Still to do:

* update [NuGet](https://www.nuget.org/packages/Knockoutjs): only @SteveSanderson has access.
* update [CDNJS](https://cdnjs.com/libraries/knockout): I'll take care of this.

**Comment by mbest** [comment]
Okay, Knockout on CDNJS is updated "automatically" from NPM, so I don't need to do anything.

**Comment by Bearcosta** [comment]
Hey!

I am eagerly waiting to start using 3.3 now. But for some reason the NuGet package doesn't haven't been updated to 3.3. When will this happen?

Cheers!

**Comment by Bearcosta** [comment]
It was apparently added to NuGet March 6th.

Cheers!

**Comment by jpsala** [comment]
Thanks a lot for your work! I enjoy my work as a programer lot more
